doc/developer/design/20210401_volatile_sources.md

```diff
@@ -0,0 +1,294 @@
+into a nonvolatile one. If we can write down the data we see as it flows into
+Materialize, then it doesn't matter if the upstream source garbage collects it.
+
+## Alternatives
+
+### Forced materialization
+
+One considered alternative was forcing all PubNub and SSE sources to be
+materialized. This neatly sidesteps the volatility within a single
+`materialized` binary, because it guarantees there is only one instantiation
+of each volatile source.
+
+This _severely_ limits Materialize's expressivity, though. One use case we
+expect to be very common in demos is to only materialize the last _n_ seconds
+of data, which requires both an unmaterialized source and a materialized view:
+
+```sql
+CREATE SOURCE market_orders_raw FROM PUBNUB SUBSCRIBE KEY 'sub-c-4377ab04-f100-11e3-bffd-02ee2ddab7fe' CHANNEL 'pubnub-market-orders';
+
+CREATE MATERIALIZED VIEW market_orders AS
+  SELECT *
+  FROM (
+      SELECT
+        (text::jsonb)->>'bid_price' AS bid_price,
+        (text::jsonb)->>'order_quantity' AS order_quantity,
+        (text::jsonb)->>'symbol' AS symbol,
+        (text::jsonb)->>'trade_type' AS trade_type,
+        ((text::jsonb)->'timestamp')::bigint * 1000 AS timestamp_ms
+      FROM market_orders_raw
+    )
+  WHERE mz_logical_timestamp() BETWEEN timestamp_ms AND timestamp_ms + 30000
+```
+
+### One-shot sources
+
+One other discussed alternative was to introduce a "one-shot" source, which
+could be _used_ in exactly one materialized view:
+
+```
+CREATE MATERIALIZED VIEW market_orders AS
+ WITH market_orders_raw AS SOURCE (
+     PUBNUB SUBSCRIBE KEY 'sub-c-4377ab04-f100-11e3-bffd-02ee2ddab7fe' CHANNEL 'pubnub-market-orders';
+  )
+  SELECT *
+  FROM (
+      SELECT
+        (text::jsonb)->>'bid_price' AS bid_price,
+        (text::jsonb)->>'order_quantity' AS order_quantity,
+        (text::jsonb)->>'symbol' AS symbol,
+        (text::jsonb)->>'trade_type' AS trade_type,
+        ((text::jsonb)->'timestamp')::bigint * 1000 AS timestamp_ms
+      FROM market_orders_raw
+    )
+  WHERE mz_logical_timestamp() BETWEEN timestamp_ms AND timestamp_ms + 30000
+```
+
+The idea here is to force you to type out the `SOURCE` definition every time you
+want to use it, to make it more clear that you will get different data in
+different views that use that source. But this design does not fit well into the
+existing catalog infrastructure, which maintains a very clear separation between
+sources and views. And even with this design, I _still_ think we'd want to
+introduce the notion of volatility, because we'll want to exactly-once sinks and
+active-active replication to know that `market_orders` is not a well-behaved
+view.
+
+### Wait for volatility removal
+
+We could, in principle, reject the idea of volatile sources, and wait for source
+persistence to unlock [volatility removal](#volatility-removal). Then the story
+for using volatile sources with Materialize would be to _require_ using them
```

Comment:
I just want to call out that in addition to the dev time this makes me nervous because I'm not at all sure how something like this would work with active-active (how do you guarantee that two different instances of MZ read from PubNub at the same time). I think doing so would require a sufficiently advanced implementation where multiple MZ instances could share what they have persisted with each other.

I think its a totally fine thing to leave in the doc though.

Reply:
ah your comment about pubnub allowing storage applies here as well I think. But I think regardless, any source that is analogous to jumping in front of a firehose of events will have this sort of problem?

Reply:
> how do you guarantee that two different instances of MZ read from PubNub at the same time

yeah, totally, seems hard